File: searchingAlgo/rabinKarpSearch/rabinKarp.py
def rabinKarp(text, pattern):
    # Define prime number to use in hashing (it's a good practice to choose a prime number)
    prime = 101

    # Calculate the length of the text and pattern
    n = len(text)
    m = len(pattern)

    # Calculate the hash value for the pattern
    pattern_hash = 0
    for i in range(m):
        pattern_hash = (pattern_hash + ord(pattern[i])) % prime

    # Calculate the hash value for the initial window of text
    text_hash = 0
    for i in range(m):
        text_hash = (text_hash + ord(text[i])) % prime

    # Iterate through the text with a sliding window
    for i in range(n - m + 1):
        # Check if the hash values match
        if pattern_hash == text_hash:
            # If hash values match, perform a character-by-character comparison
            match = True
            for j in range(m):
                if pattern[j] != text[i + j]:
                    match = False
                    break
            if match:
                # Pattern found at index 'i'
                return i

        # Update the hash value for the next window of text
        if i < n - m:
            # Remove the leftmost character and add the rightmost character
            text_hash = (text_hash - ord(text[i]) + ord(text[i + m])) % prime

            # Ensure the hash value is non-negative
            if text_hash < 0:
                text_hash += prime

    # Pattern not found in the text
    return -1

# Test the algorithm
text = "AABBACHADA"
pattern = "BBAC"
index = rabinKarp(text, pattern)
if index != -1:
    print(f"Pattern found at index: {index+1}")
else:
    print("Pattern not found")

File: searchingAlgo/rabinKarpSearch/rabinKarp.c
#include<stdio.h>
#include<string.h>
 

int rabinKarp(char txt[], char pat[], int q) 
{ 
	int txt_len = strlen(txt);
	int pat_len = strlen(pat);
    int d = 256;
	int p=0,t=0,h=1;
	
	for(int i=0;i<pat_len-1;i++)
	    h = (h*d)%q;
	
	for(int i=0;i<pat_len;i++)
	{
	    p = (d*p + pat[i])%q;
	    t = (d*t + txt[i])%q;
	}
	for(int i=0;i<=txt_len-pat_len;i++)
	{
	    if(p == t)
	    {
	        int j;
	        for(j=0;j<pat_len;j++)
	        {
	            if(txt[i+j] != pat[j])
	                break;
	        }
	        if(j == pat_len)
	            return i;
	    }
	    if(i < txt_len-pat_len)
	    {
	        t = ((t - txt[i]*h)*d + txt[pat_len+i])%q;
	        if(t < 0)
	            t = t + q;
	    }
	}
	return -1;
} 

int main()
{
    char text[] = "AABBACHADA";
    char pat[] = "BBAC";
    int index =  rabinKarp(text,pat,101);
    if(index != -1)
        printf("Pattern found at: %d\n", index+1);
    else
        printf("Pattern not found\n");
    
}
